### ISSUE TITLE
Paging large files is slow

### ISSUE THREAD

**Issue by alex-r-bigelow** [Kitware/candela#296]
Specifically, `getHistograms()` doesn't make use of `limit` and `offset` for mongo databases (mongodb's mapreduce doesn't support an offset setting)

**Comment by jeffbaumes** [comment]
You could talk to @manthey about this and what possible strategies might be.

One idea: as soon as data starts loading up can you start making histograms for all pages in the background and cache the results? This would only need one pass to get them all. This would enable faster paging but of course changing the filter or page size would require recomputing all of that and probably make those updates even slower than before.

**Comment by alex-r-bigelow** [comment]
Yeah, we're actually already sort of doing this—all the histogram computations are launched in parallel as soon as the dataset schema is available (if they're not already cached).

The page histograms are the only ones that aren't cached, because they *should* generally involve faster passes than the others because they can terminate early. Though now that I think about it, the last page will require a full table scan anyway, so maybe we should turn on caching for page histograms as well.

I think it's already the case that flat files terminate early for the page histogram computation (TODO: I should test this)... if we cache them, we could probably even store byte offsets for each page to help us fast-forward to the right spot (precluding the need for a full table scan).

This issue, however, is a little simpler: at the moment, mongo page histogram calculations still do full table scans. I think there's a way to give a counter to the map function to prevent emitting items on different pages, but it's proving a little messier than I anticipated.

**Comment by jeffbaumes** [comment]
Another thought is that we could lower priority on this one for now. Function trumps performance at this stage.

**Comment by alex-r-bigelow** [comment]
This is partially fixed; mongo doesn't have an offset parameter, so later pages will still be slow (we keep a counter internally in the mapreduce code to simulate an offset).
